daiquiri/__init__.py

```diff
@@ -51,21 +51,66 @@ def process(self, msg, kwargs):
         return msg, kwargs
 
 
+PARAMETER_TEMPLATE = ' [{0}: {1}]'
+
+
+class ArbitraryContextAdapter(logging.LoggerAdapter):
+    """Logger adapter to add context string to log record's extra data
+
+    Keywords passed to the log call are formatted into a context
+    string, which is then added to the "extra" dictionary passed to
+    the underlying logger so they are emitted with the log message and
+    available to the format string.
+
+    Example:
+      A formatter with the format string of "%(context)s %(messages)"
+      called like logger.info("A message.", test1="a", test2="b")
+      would log the message: " [test1: a] [test2: b] A message."
+
+    Special keywords:
+
+    extra
+      An existing dictionary of extra values to be passed to the
+      logger. If present, the dictionary is copied and extended.
+    """
+
+    def process(self, msg, kwargs):
+        # Make a new extra dictionary combining the values we were
+        # given when we were constructed and a 'context' key which
+        # will contain a formatted string containing anything from
+        # kwargs.
+        extra = self.extra.copy()
+        if 'extra' in kwargs:
+            extra.update(kwargs.pop('extra'))
+        # Format any unknown keyword arguments into the context string.
+        context = ''
+        for name in list(kwargs.keys()):
+            if name == 'exc_info':
+                continue
+            context += PARAMETER_TEMPLATE.format(name, kwargs.pop(name))
+        extra['context'] = context
+        extra['_daiquiri_extra'] = extra
+        kwargs['extra'] = extra
+        return msg, kwargs
```

Comment:
There's a lot of duplication here with `KeywordArgumentAdapter` so this clearly should be factored with some inheritance.

If I understand correctly the only thing added here is `extra['context']` so I wonder to which extent this could not be a built-in feature of `KeywordArgumentAdapter` directly?

Reply:
I like that! It should probably be in a different class than `ColorFormatter`, just so users can pick either `ColorFormatter` or that new class, let's say `ExtraFormatter`.
And we can have a third class combining the two that should be named `ColorExtraFormatter` and used by default. I'd include `rest-of-extra` in the default context string.

With a better name, such as, I don't know, `%(extras)`? :)